src/data/dem_data.js

```diff
@@ -0,0 +1,170 @@
+// @flow
+const {RGBAImage} = require('../util/image');
+
+export type SerializedDEMData = {
+    uid: string,
+    scale: number,
+    levels: Array<ArrayBuffer>
+};
+
+class Level {
+    width: number;
+    height: number;
+    border: number;
+    stride: number;
+    data: Int32Array;
+
+    constructor(width: number, height: number, border: number, data: ?Int32Array) {
+        if (height <= 0 || width <= 0) throw new RangeError('Level must have positive height and width');
+        this.width = width;
+        this.height = height;
+        this.border = border;
+        this.stride = this.width + 2 * this.border;
+        this.data = data || new Int32Array((this.width + 2 * this.border) * (this.height + 2 * this.border));
+    }
+
+    set(x: number, y: number, value: number) {
+        this.data[this.idx(x, y)] = value + 65536;
+    }
+
+    get(x: number, y: number) {
+        return this.data[this.idx(x, y)] - 65536;
+    }
+
+    idx(x: number, y: number) {
+        if (x < -this.border || x >= this.width + this.border ||  y < -this.border || y >= this.height + this.border) throw new RangeError('out of range source coordinates for DEM data');
+        return (y + this.border) * this.stride + (x + this.border);
+    }
+}
+
+class DEMData {
+    uid: string;
+    scale: number;
+    data: Array<Level>;
+    loaded: boolean;
+
+    static deserialize(serializedData: SerializedDEMData) {
+        const structdata = new Int32Array(serializedData.levels[0]);
+        const stride = Math.sqrt(structdata.length);
+        const data = [new Level(stride / 2, stride / 2, stride / 4, structdata)];
+        return new DEMData(serializedData.uid, serializedData.scale, data);
+    }
+
+    constructor(uid: string, scale: ?number, data: ?Array<Level>) {
+        this.uid = uid;
+        this.scale = scale || 1;
+        this.data = data ? data : [];
+        this.loaded = !!data;
+    }
+
+    loadFromImage(data: RGBAImage) {
+        // Build level 0
+        this.data = [new Level(data.width, data.height, data.width / 2)];
```

Comment:
data.width and data.height should always be the same, and the border is needed because texture dimensions have to be power of 2, so we put a border of 1/2 the original dimension on each side to effectively double the height and width. this should be clearer.

Reply:
Ah, interesting!